File: kotlin/binary-search/src/main/kotlin/BinarySearch.kt
object BinarySearch {
    fun search(input: List<Int>, value: Int): Int = binarySearch(input, value,0, input.size - 1)

    private tailrec fun binarySearch(input: List<Int>, value: Int, lower: Int, upper: Int): Int {
        val middleIndex = upper + lower / 2

        return when {
            lower > upper -> -1
            input[middleIndex] == value -> middleIndex
            input[middleIndex] < value -> binarySearch(input, value, middleIndex + 1, upper)
            else -> binarySearch(input, value, lower, middleIndex - 1)
        }
    }
}

File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find (input: 'T array) value =
    let rec aux min max = 
        if min > max then
            None
        else
            let middle = (min + max) / 2            
            match input.[middle] with
            | x when x < value -> aux (middle + 1) max
            | x when x > value -> aux min (middle - 1)
            | _ -> Some middle
            
    aux 0 (input.Length - 1)